department-of-veterans-affairs/abd-vro
# Adds BGS response fields to response to fit new RabbitMQ strategy

## What was the problem?

Current implementation of `svc-bgs-api` does not return responses as expected according to the [VRO RabbitMQ Strategy](https://github.com/department-of-veterans-affairs/abd-vro/wiki/VRO-RabbitMQ-Strategy).

Associated tickets or Slack threads:
- #2300 
- Follow on for 
  - #2176 

## How does this fix it?[^1]

changes the responses of `svc-bgs-api` to the format outlined in the [VRO RabbitMQ Strategy](https://github.com/department-of-veterans-affairs/abd-vro/wiki/VRO-RabbitMQ-Strategy).

## How to test this PR
Build images and run docker containers 
```
source scripts/setenv.sh
export COMPOSE_PROFILES="bgs"
./gradlew docker
./gradlew -p mocks docker
./gradlew :dockerComposeUp
./gradlew :app:dockerComposeUp
./gradlew -p mocks :dockerComposeUp
```
2. Verify RabbitMQ, `svc-bgs-api`, and `mock-bgs-api` are running and ready to receive traffic. Note `mock-bgs-api` can take about minute.
3. Create a new queue on the `bgs-api` exchange called `response`. Be sure to bind the queue to the exchange. 

### Valid Request: 
1. Open to the rabbitmq console to the [add-note](http://localhost:15672/#/queues/%2F/getClaimContentionsQueue) publish a message with the body `{"vbmsClaimId":111,"claimNotes":["hey"]}` and add a message property called `reply_to` with the value `response`.
2. Open the [response queue page](http://localhost:15672/#/queues/%2F/response) and perform the get message action
verify the response is:
```
{
  "statusCode": 200,
  "statusMessage": "OK"
}
```

### Invalid Request - Missing `claimNotes` field: 
1. Open to the rabbitmq console to the [add-note](http://localhost:15672/#/queues/%2F/getClaimContentionsQueue) publish a message with the body `{"vbmsClaimId":111,}` and add a message property called `reply_to` with the value `response`.
2. Open the [response queue page](http://localhost:15672/#/queues/%2F/response) and perform the get message action
verify the response is:

```
{
  "statusCode": 400,
  "statusMessage": "BAD_REQUEST",
  "messages": [
    {
      "key": "ArgumentError",
      "severity": "ERROR",
      "status": 400,
      "text": "missing claimNotes or veteranNote",
      "timestamp": "2023-12-05T22:29:12+00:00",
      "httpStatus": "BAD_REQUEST"
    }
  ]
}
```

### Invalid Request - Missing `vbmsClaimId` field: 
1. Open to the rabbitmq console to the [add-note](http://localhost:15672/#/queues/%2F/getClaimContentionsQueue) publish a message with the body `{"claimNotes":["hey"]}` and add a message property called `reply_to` with the value `response`.
2. Open the [response queue page](http://localhost:15672/#/queues/%2F/response) and perform the get message action
verify the response is:

```
{
  "statusCode": 400,
  "statusMessage": "BAD_REQUEST",
  "messages": [
    {
      "key": "ArgumentError",
      "severity": "ERROR",
      "status": 400,
      "text": "vbmsClaimId is required for claimNotes",
      "timestamp": "2023-12-05T22:33:05+00:00",
      "httpStatus": "BAD_REQUEST"
    }
  ]
}
```

### Downstream `svc-bgs-api` internal server error 
1. Shutdown `mock-bgs-api` container.
2. Open to the rabbitmq console to the [add-note](http://localhost:15672/#/queues/%2F/getClaimContentionsQueue) publish a message with the body `{"claimNotes":["hey"]}` and add a message property called `reply_to` with the value `response`.
3. Open the [response queue page](http://localhost:15672/#/queues/%2F/response) and perform the get message action
verify the response is:

```
{
  "statusCode": 500,
  "statusMessage": "INTERNAL_SERVER_ERROR",
  "messages": [
    {
      "key": "SocketError",
      "severity": "ERROR",
      "status": 500,
      "text": "getaddrinfo: Name does not resolve (mock-bgs-api:8080)",
      "timestamp": "2023-12-05T22:35:27+00:00",
      "httpStatus": "INTERNAL_SERVER_ERROR"
    }
  ]
}

```


[^1]: [Pull-Requests guidelines](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Pull-Requests). If PR is significant, update [Current Software State](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Current-Software-State) wiki page.
[^secrel]: To check if a PR will succeed in the SecRel workflow, [test PRs in the SecRel pipeline](https://github.com/department-of-veterans-affairs/abd-vro-internal/wiki/Secure-Release-process#to-test-prs-in-the-secrel-pipeline).
